### PR TITLE
Prevent meaningless enqueue before SyncAll

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
this is the second half, based on #33247 to simplify review.
Prevent task enqueue before we SyncAll, this is to prevent double handling resources during bootstrap. 

In my test env, the istio ready time before is
```
2021-06-03T12:05:00.134841Z     error   ====waitForCacheSync: committed 28181, expected 24816
2021-06-03T12:05:00.134877Z     info    ads     All caches have been synced up in 31.57432894s, marking server ready
```
with this patch: 
```
2021-06-03T12:13:19.954575Z     error   ====waitForCacheSync: committed 18512, expected 18512
2021-06-03T12:13:19.954612Z     info    ads     All caches have been synced up in 22.187328946s, marking server ready
```

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/bootstrap/server.go
Comment (howardjohn) [inline pilot/pkg/bootstrap/server.go:806]: should either remove entirely or drop the level and `===`

File: pilot/pkg/config/kube/crdclient/client.go
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/client.go:187]: Why is it better to skip updates then SyncAll here? I get for the k8s controller, but not or crd controller.. we still have the same amount of work to do?

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:318]: Is the requiring *only* that the event is in the informer cache, or that its in the informer cache *AND* been processed? We have some internal indexes like `servicesMap` and pod mapping that will not be up to date?

**Review by howardjohn — CHANGES_REQUESTED** [review]
Looks like it could be a good improvement!

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (stevenctl) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:275]: nit: can you make this clearer by adding "when calling SyncAll, but before SyncAll has completed"

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:318]: We won't actually run the queue until `initialSync` is loaded. We just start enqueueing while `SyncAll` is running. 

I'm a bit concerned about a scenario where the "ignore" condition holds and we completely drop some events not captured in `SyncAll`. I don't _think_ it's possible, if the informer indexer List calls are populated from the same source as the events.

**Comment by stevenctl** [comment]
Seems like a good improvement. One question I have though – we skip enqueuing because we assume the event will already be handled explicitly in `SyncAll`. What if `SyncAll` ignores a specific type. 

Our k8s controller handles node, service, pod, endpoint. If some other kind comes along and needs to be handled, we'd totally drop some initial events for it? Can we have a list of `SyncAllKinds` and only ignore events in that set?

**Comment by hzxuzhonghu** [comment]
We will and should SyncAll for all kinds of resources

File: pilot/pkg/bootstrap/server.go
Comment (hzxuzhonghu) [reply pilot/pkg/bootstrap/server.go:806]: Yes, see  #33247, once that is merged, i will remove the commits in front

File: pilot/pkg/config/kube/crdclient/client.go
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crdclient/client.go:187]: Without `SyncAll ` here, let's take an example here:

If we have total 100 configs here, without syncAll, let's say 100 configs enqueued and but only 50 is added to inboundedUpdate, and the other 50 has not been run. Then the time we mark istiod ready is not right, in this case, sidecar could get partial xds, which is dangerous.

Comment (howardjohn) [reply pilot/pkg/config/kube/crdclient/client.go:187]: can we just enqueue some signal marks us as ready? then we know once we have processed that all 100 configs were processed as well. I think we do that in other places. That seems a bit simpler

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:318]: Right, before we sync all, ignore all the events.

> I don't think it's possible, if the informer indexer List calls are populated from the same source as the events.
Yes, could not skip any resource event

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:275]: >but before SyncAll has completed

This is not right, do not modify it after SyncAll

File: pilot/pkg/config/kube/crdclient/client.go
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crdclient/client.go:187]: This will require the queue to be stateful.  And there is a pr replacing the queue with k8s rate limiting queue.  I am afraid this is not a good idea

Comment (howardjohn) [reply pilot/pkg/config/kube/crdclient/client.go:187]: I am not sure I understand the issue. Regardless of the queue implementation, its always ordered. So when we dequeue a "ready" signal we can mark ourselves ready?

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (stevenctl) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:279]:
```go
func (cl *Client) SyncAll() {	
   cl.beginInitialSync.Store(true)
```

Seems to me like it says "we're in the process of syncing, but it may or may not have completed".

**Comment by stevenctl** [comment]
https://github.com/istio/istio/pull/33257#issuecomment-854288208

We _should_ but its the sort of thing that doesn't have any easy check and could cause obscure bugs.

File: pilot/pkg/config/kube/crdclient/client.go
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crdclient/client.go]: I think i understand what you mean, but the tricky part is that will require a lock because SyncAll and other resource event handler can enqueue some tasks concurrently. That seems super hard to do , here the solution maybe do some little percentage of duplicate event handling, but from the sync view, it is the easiest.

**Comment by hzxuzhonghu** [comment]
Make some sense, I will add `SyncAllKinds` for serviceController

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go]: changed to `beginSync`, i think it is more clear. Indicates the controller has begun syncing resources.

**Comment by stevenctl** [comment]
This looks good to me, I'll let @howardjohn re-review before approving.

**Comment by hzxuzhonghu** [comment]
@howardjohn Which one should we move in?

File: pilot/pkg/config/kube/crdclient/cache_handler.go
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/cache_handler.go:91]: nit: incrementEvent after the `if`?

File: pilot/pkg/config/kube/crdclient/client.go
Comment (howardjohn) [inline pilot/pkg/config/kube/crdclient/client.go:210]: related to previous PR: why do we parallelized cl.kinds? Isn't it 100% CPU bound? does it actually speed things up?

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/controller.go:361]: why not do this at start of the function?

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:361]: Also this function is not called on any pod, just ones that are in needResync, so I am not sure its actually something we want to skip

**Review by howardjohn — APPROVED** [review]
TBH I am very concerned that this is so complex and has no test coverage, and any error here can result in an outage. We need to clean this up in a followup.

File: pilot/pkg/config/kube/crdclient/cache_handler.go
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crdclient/cache_handler.go:91]: I understand this is to increase for every resource event, no matter whether it triggers push. So i put it before

File: pilot/pkg/config/kube/crdclient/client.go
Comment (hzxuzhonghu) [reply pilot/pkg/config/kube/crdclient/client.go:210]: There are many cpu cores,  and the go routines scheduling take charge of mapping them to kernel thread and then scheduled by kernel. The cpu can exceed 100%, but it does help speed up.

File: pilot/pkg/serviceregistry/kube/controller/controller.go
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/controller.go:361]: This is what i thought: in k8s, for normal endpoint and pods, the pod creation must be before the endpoints, endpoints-controller list-watch pods and then reconciles the endpoints. It must be that during startup, istiod list-watch pods/endpoints, after cache syncs, for all the endpoints we can find the related pods. 

The issue why endpoint event occur before pods, is that there are two streams on flying.

**Pull request merged**
